Test for dyadic rationals reports error

This error happens stochastically, about once every five to ten test run. The failing test is `amount::dyadic_rationals::dyadic_rationals_tests::add_sub_pbt`.

```
failures:

---- amount::dyadic_rationals::dyadic_rationals_tests::add_sub_pbt stdout ----
thread 'amount::dyadic_rationals::dyadic_rationals_tests::add_sub_pbt' panicked at twenty-first/src/amount/dyadic_rationals.rs:155:9:
Right-hand-side argument cannot be larger than left-hand-side for subtraction
stack backtrace:
   0: rust_begin_unwind
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/std/src/panicking.rs:595:5
   1: core::panicking::panic_fmt
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/panicking.rs:67:14
   2: <twenty_first::amount::dyadic_rationals::DyadicRational as core::ops::arith::Sub>::sub
             at ./src/amount/dyadic_rationals.rs:155:9
   3: twenty_first::amount::dyadic_rationals::dyadic_rationals_tests::add_sub_pbt
             at ./src/amount/dyadic_rationals.rs:341:37
   4: twenty_first::amount::dyadic_rationals::dyadic_rationals_tests::add_sub_pbt::{{closure}}
             at ./src/amount/dyadic_rationals.rs:336:22
   5: core::ops::function::FnOnce::call_once
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/ops/function.rs:250:5
   6: core::ops::function::FnOnce::call_once
             at /rustc/cc66ad468955717ab92600c770da8c1601a4ff33/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    amount::dyadic_rationals::dyadic_rationals_tests::add_sub_pbt

test result: FAILED. 407 passed; 1 failed; 4 ignored; 0 measured; 0 filtered out; finished in 7.51s

error: test failed, to rerun pass `-p twenty-first --lib`
make: *** [Makefile:63: test] Error 101
```

--- a/twenty-first/src/amount/dyadic_rationals.rs
+++ b/twenty-first/src/amount/dyadic_rationals.rs
@@ -135,13 +135,15 @@ impl Ord for DyadicRational {
             return other.cmp(self).reverse();
         }
 
+        // We now know that: self.exponent >= other.exponent
+
         // Put on same denominator
         let max_exponent = self.exponent;
         let min_exponent = other.exponent;
-        let shift = max_exponent - min_exponent + 1;
+        let shift = max_exponent - min_exponent;
 
         // Add numerators
-        let shifted_other_mantissa = shift * other.mantissa.clone();
+        let shifted_other_mantissa = other.mantissa.clone() << shift;
 
         self.mantissa.cmp(&shifted_other_mantissa)
     }
@@ -154,7 +156,7 @@ impl Sub for DyadicRational {
         // panic if rhs is larger
         assert!(
             self >= rhs,
-            "Right-hand-side argument cannot be larger than left-hand-side for subtraction"
+            "Right-hand-side argument cannot be larger than left-hand-side for subtraction.\nLHS was: {self:?},\nRHS: {rhs:?}"
         );
 
         // Put on same denominator
@@ -362,6 +364,32 @@ mod dyadic_rationals_tests {
         };
 
         assert_ne!(a, c);
+    }
+
+    #[test]
+    fn ordering_unit_test() {
+        let a: DyadicRational = 16.into();
+        let b = DyadicRational {
+            exponent: 2,
+            mantissa: 64u32.into(),
+        };
+
+        assert!(!(a > b));
+        assert!(!(b > a));
+        assert!(!(a < b));
+        assert!(!(b < a));
+        assert!(a <= b);
+        assert!(a >= b);
+
+        let c = DyadicRational {
+            exponent: 3,
+            mantissa: 127u32.into(),
+        };
+
+        assert!(b > c);
+        assert!(c < b);
+        assert!(a > c);
+        assert!(c < a);
     }
 
     #[test]

